server/src/main/java/org/opensearch/search/aggregations/AggregationCollectorManager.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.aggregations;
+
+import org.apache.lucene.search.Collector;
+import org.apache.lucene.search.CollectorManager;
+import org.opensearch.search.internal.SearchContext;
+import org.opensearch.search.query.ReduceableSearchResult;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+
+import static org.opensearch.search.aggregations.DefaultAggregationProcessor.createCollector;
+
+/**
+ * {@link CollectorManager} to take care of aggregation operators both in case of concurrent and non-concurrent
+ * segment search
+ */
+public class AggregationCollectorManager implements CollectorManager<Collector, ReduceableSearchResult> {
+    private final SearchContext context;
+
+    public AggregationCollectorManager(SearchContext context) {
+        this.context = context;
+    }
+
+    @Override
+    public Collector newCollector() throws IOException {
+        List<Aggregator> nonGlobalAggregators = context.aggregations().factories().createTopLevelNonGlobalAggregators(context);
+        assert !nonGlobalAggregators.isEmpty() : "Expected atleast one non global aggregator to be present";
+        context.aggregations().addNonGlobalAggregators(nonGlobalAggregators);
```

Comment:
I believe we under-use the collector manager here and still rely on search context:
 - why the `reduce` is empty but it could (and should) reduce over the aggregations? 
 - why we need to store all aggregations in context? (whereas reducing the result in the collector manager would eliminate the need to do that)

Reply:
1. `reduce` in case of Aggregation is performed using the [InternalAggregations.reduce](https://github.com/opensearch-project/OpenSearch/pull/7514/files#diff-28d65e7be4bf91608a1b02e9745afc3765236b2484fe6b4da487440500930fadR33). This works on intermediate representation of Aggregations which is `InternalAggregation`. Building `InternalAggregation` can be done via the `Aggregator` interface methods whereas `reduce` in `CollectorManager` works only using `Collectors`. The Aggs during collection phase can be of type `BucketCollector` or `MultiBucketCollector` which doesn't allow to use `Aggregators` explicitly which is needed to create `InternalAggregation` via `buildTopLevel` method.

2. For above reason we are storing the aggregations in context which is cleaned up in `postProcess`. This logic of storing `Aggregations` in context exists even today in existing implementation.

3. Also this way we are able to reuse the code in `postProcess` both for `DefaultAggregationProcessor` and `ConcurrentAggregationProcessor` implementation.

Reply:
It was done mainly to reuse the code between Concurrent and Default `AggregationProcessor`. Doing it in `reduce` was not turning out to be a cleaner way.
* `AggregationCollectorManager` is being used both during concurrent and non-concurrent search path. In both the cases, it will be used to create the collector instances. So all the aggregators created as part of `newCollector` will be tracked in `SearchContextAggregations` irrespective of where `reduce` is performed
*  It require conversions between `Collector` to `BucketCollector` to `Aggregator`. 
*  The `QueryResult` which will be created with `reduce` will contain only results for non global aggregations. Then later in `AggregationProcessor::postProcess` global aggregations will be computed and that now needs to be added in the `QueryResults`. Whereas performing `reduce` later in `postProcess` will perform it for both global and non-global aggregation and update the query results with final output.

> This logic of storing Aggregations in context exists even today in existing implementation.

>> This is not the same: we used to have single collector with N aggregators, now we end up with M collectors, each with N aggregators.

What I mean is using `context` object to keep track of all the aggregators to be used later in `postProcess` stage. Yes now instead of N aggs references it will have `sliceCount * N aggs` references which should still be fine.

I think this refactor of moving the reduction into the `AggregationCollectorManager::reduce` will help when we deprecate the `DefaultAggregationProcessor`. Thoughts ?

Reply:
@reta If you are fine, we can revisit this after adding the change to dynamically control the concurrent search vs non concurrent search path. I will create a draft PR with changes in `AggregatorCollectorManager::reduce` and we can discuss if we want to move forward with that or keep it as is. Thoughts ?

Reply:
@sohami working on prototype (based of your pull request), will get back shortly.

Reply:
I have done pretty quick changes in AggregationCollectorManager  by implementing the proper `reduce` method, and it allowed me:
 - pass global aggregators to collector manager (so they could be used in reduce phase)
 - cleanup the search context by removing global/non-global aggregators from it (collector manager takes care of it)
 - cleanup the aggregation phase by removing the result processing (collector manager takes care of it)
 - for non-concurrent aggregations, the collector manager's `reduce` is called manually on preProcess callback (this is the reuse across concurrent and non-concurrent flows)

There are few minor things to address, I just didn't have enough time to finish them today (I have 12 test cases failing on `:rest-api-spec:yamlRestTest` related to globals I beleive), but I have sent this draft pull request [1] if you are curious what those changes are.

[1] https://github.com/sohami/OpenSearch/pull/27

Reply:
@reta Thanks for sharing the draft. 
* I was trying to keep the global aggregation processing separate from the `searcher.search()` phase call in query phase to keep the current flow. My concern was in case of any failure if that will cause some regression in functionality in current flow. But on looking more, it seems like it should be fine as any failure during Global Agg processing also end up with `QueryPhaseExecutionException`.  
* I like that we are now keeping all `collectors` related info in `CollectorManager` instead of `SearchContext`. 
* Calling `reduce` in `postProcess` callback will have issues in concurrent case, as it will be called twice - one by searcher and other by this callback. But we can solve that by overwriting the `postProcess` in `ConcurrentProcessor`. In current implementation this was solved by making `reduce` as no-op instead.
* However we will also make Global Aggregation processing to use concurrent flow as well. So now when reduce phase of non-global and global aggs are called they will populate the `QueryResult` separately. So we will need a mechanism to merge those before returning the results.  In current implementation, with `reduce` being no-op I was thinking of reusing the `SearchContext` to keep the `postProcess` logic common. But with `reduce` phase mechanism we will need probably some special handling.

I will work on the changes (and will also make Global Aggs handling concurrent) based on your prototype. Basically will keep the idea of storing collectors in `CollectorManager` instead of `SearchContext` and see how that looks like.

Reply:
Thanks @sohami 

> Calling reduce in postProcess callback will have issues in concurrent case, as it will be called twice - one by searcher and other by this callback. But we can solve that by overwriting the postProcess in 

No, it won't be called twice, it only called when query context won't have aggregators, with concurrent flow the aggregators will be set in reduce phase (there is condition in the begging of the method)

Reply:
> @reta Have updated the PR to use `CollectorManager` and also added the support for `GlobalAggregation` with CollectorManager

Thanks @sohami, it looks much better (to me), thank you.